DataFormats/PortableTestObjects/src/classes_def.xml

```diff
@@ -1,24 +1,16 @@
 <lcgdict>
   <class name="portabletest::TestSoA"/>
```

Comment:
Isn't `TestSoA` a duplicate too (being an alias to `portabletest::TestSoALayout<128,false>`)?

Maybe move all the `TestSoALayout<...>` at the beginning of the file?

Reply:
Actually I would prefer to keep the aliases and remove the template specialisations.

Reply:
I agree using the alias would be fine. Unfortunately, I re-tested this and it only works up to 2 layouts, but with the 3 layouts version, the dictionary generation crashes:
```
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02828/el8_amd64_gcc12/lcg/root/6.30.05-00c2425f5a7c62d1e0d7747537277755/bin/rootcling -reflex -f tmp/el8_amd64_gcc12/src/DataFormats/PortableTestObjects/src/DataFormatsPortableTestObjects/a/DataFormatsPortableTestObjects_xr.cc -inlineInputHeader -failOnWarnings -rmf tmp/el8_amd64_gcc12/src/DataFormats/PortableTestObjects/src/DataFormatsPortableTestObjects/a/DataFormatsPortableTestObjects_xr.rootmap -rml libDataFormatsPortableTestObjects.so -DALPAKA_DEFAULT_HOST_MEMORY_ALIGNMENT=128 -DALPAKA_DISABLE_VENDOR_RNG -DCMS_DICT_IMPL -D_REENTRANT -DGNUSOURCE -D__STRICT_ANSI__ -DGNU_GCC -D_GNU_SOURCE -DEIGEN_DONT_PARALLELIZE -DEIGEN_MAX_ALIGN_BYTES=64 -DTBB_USE_GLIBCXX_VERSION=120301 -DTBB_SUPPRESS_DEPRECATED_MESSAGES -DTBB_PREVIEW_RESUMABLE_TASKS=1 -DTBB_PREVIEW_TASK_GROUP_EXTENSIONS=1 -DBOOST_SPIRIT_THREADSAFE -DPHOENIX_THREADSAFE -DBOOST_MATH_DISABLE_STD_FPCLASSIFY -DBOOST_UUID_RANDOM_PROVIDER_FORCE_POSIX -DCMSSW_GIT_HASH="CMSSW_14_1_X_2024-03-11-1100" -DPROJECT_NAME="CMSSW" -DPROJECT_VERSION="CMSSW_14_1_X_2024-03-11-1100" -I/data/user/cano/ZVertexMultiLayout/CMSSW_14_1_X_2024-03-11-1100/src -I/data/user/cano/ZVertexMultiLayout/CMSSW_14_1_X_2024-03-11-1100/poison -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02828/el8_amd64_gcc12/cms/cmssw-patch/CMSSW_14_1_X_2024-03-11-1100/src -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02828/el8_amd64_gcc12/external/alpaka/1.1.0-c6af69ddd6f2ee5be4f2b069590bae19/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02828/el8_amd64_gcc12/external/pcre/8.43-37eb2e8b73bab83d6645ecfd5d73dcaa/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02828/el8_amd64_gcc12/external/boost/1.80.0-fe2842446e7b8c2042e86925ee67be29/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02828/el8_amd64_gcc12/external/bz2lib/1.0.6-d065ccd79984efc6d4660f410e4c81de/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02828/el8_amd64_gcc12/external/libuuid/2.34-27ce4c3579b5b1de2808ea9c4cd8ed29/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02828/el8_amd64_gcc12/lcg/root/6.30.05-00c2425f5a7c62d1e0d7747537277755/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02828/el8_amd64_gcc12/external/tbb/v2021.9.0-c3903c50b52342174dbd3a52854a6e6d/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02828/el8_amd64_gcc12/external/xz/5.2.5-6f3f49b07db84e10c9be594a1176c114/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02828/el8_amd64_gcc12/external/zlib/1.2.11-51072030b7f93c3ac6c4235f21e413cb/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02828/el8_amd64_gcc12/external/eigen/3bb6a48d8c171cf20b5f8e48bfb4e424fbd4f79e-cf66836cbd6ad8d696d804181e36d61c/include/eigen3 -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02828/el8_amd64_gcc12/external/fmt/8.0.1-54e94b39f5cf29341bb9c4765764e1ca/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02828/el8_amd64_gcc12/external/md5/1.0.0-5b594b264e04ae51e893b1d69a797ec6/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02828/el8_amd64_gcc12/external/tinyxml2/6.2.0-d17873b4d6a42a43226cf689f82ec1ef/include -DCMSSW_REFLEX_DICT src/DataFormats/PortableTestObjects/src/classes.h src/DataFormats/PortableTestObjects/src/classes_def.xml
 #0 0x00007f21f0f7dc44 PrintStackTraceSignalHandler(void*) Signals.cpp:0:0
 #1 0x00007f21f0f7b45b SignalHandler(int) Signals.cpp:0:0
 #2 0x00007f21ec94ccf0 __restore_rt (/lib64/libpthread.so.0+0x12cf0)
 #3 0x00007f21ec7baaec TProtoClass::Delete(char const*) (/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02828/el8_amd64_gcc12/lcg/root/6.30.05-00c2425f5a7c62d1e0d7747537277755/bin/../lib/libCore.so+0x32caec)
 #4 0x00007f21ec7babf7 TProtoClass::~TProtoClass() (/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02828/el8_amd64_gcc12/lcg/root/6.30.05-00c2425f5a7c62d1e0d7747537277755/bin/../lib/libCore.so+0x32cbf7)
 #5 0x00007f21ec7bacc9 TProtoClass::~TProtoClass() (/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02828/el8_amd64_gcc12/lcg/root/6.30.05-00c2425f5a7c62d1e0d7747537277755/bin/../lib/libCore.so+0x32ccc9)
 #6 0x00007f21ec75129c TObjArray::Delete(char const*) (/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02828/el8_amd64_gcc12/lcg/root/6.30.05-00c2425f5a7c62d1e0d7747537277755/bin/../lib/libCore.so+0x2c329c)
 #7 0x00007f21f243074e CloseStreamerInfoROOTFile (/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02828/el8_amd64_gcc12/lcg/root/6.30.05-00c2425f5a7c62d1e0d7747537277755/bin/../lib/libRIO.so+0xca74e)
 #8 0x00007f21ed594344 FinalizeStreamerInfoWriting(cling::Interpreter&, bool) (/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02828/el8_amd64_gcc12/lcg/root/6.30.05-00c2425f5a7c62d1e0d7747537277755/bin/../lib/libCling.so+0xa3a344)
 #9 0x00007f21ed5a7f24 GenerateFullDict(std::ostream&, cling::Interpreter&, RScanner&, std::__cxx11::list<ROOT::TMetaUtils::RConstructorType, std::allocator<ROOT::TMetaUtils::RConstructorType>> const&, bool, bool, bool, bool) (/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02828/el8_amd64_gcc12/lcg/root/6.30.05-00c2425f5a7c62d1e0d7747537277755/bin/../lib/libCling.so+0xa4df24)
#10 0x00007f21ed5b0785 RootClingMain(int, char**, bool) (/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02828/el8_amd64_gcc12/lcg/root/6.30.05-00c2425f5a7c62d1e0d7747537277755/bin/../lib/libCling.so+0xa56785)
#11 0x00007f21ed5b407c ROOT_rootcling_Driver (/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02828/el8_amd64_gcc12/lcg/root/6.30.05-00c2425f5a7c62d1e0d7747537277755/bin/../lib/libCling.so+0xa5a07c)
#12 0x000000000040115e main (/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02828/el8_amd64_gcc12/lcg/root/6.30.05-00c2425f5a7c62d1e0d7747537277755/bin/rootcling+0x40115e)
#13 0x00007f21ebb5fd85 __libc_start_main (/lib64/libc.so.6+0x3ad85)
#14 0x00000000004011ae _start (/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02828/el8_amd64_gcc12/lcg/root/6.30.05-00c2425f5a7c62d1e0d7747537277755/bin/rootcling+0x4011ae)
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace.
```
Mixing the alias and developed versions of the template works fine (i.e. with `writer.pl` - `reader.pl`), but I prefer using just the developed version to avoid confusion (this is not too cumbersome when using the helper script `DataFormats/Portable/scripts/portableHostCollectionHints`  to generate the XML).

Reply:
@makortel: indeed this is duplicate. All is fine without this line (pushed).

Reply:
> Unfortunately, I re-tested this and it only works up to 2 layouts, but with the 3 layouts version, the dictionary generation crashes:

This crash might be worth of further investigation (in a form of a new CMSSW issue with instructions how to reproduce, where we'd involve ROOT folks).

Reply:
Sure, done in https://github.com/cms-sw/cmssw/issues/44386.